files/galaxy/tpv/tool_defaults.yml

```
@@ -29,6 +29,17 @@ tools:
           entity.params['requirements'] = '(GalaxyGroup == "compute") || (%s)' % training_expr if training_expr else '(GalaxyGroup == "compute")'
           entity.params['+Group'] = training_labels
           entity.params['accounting_group_user'] = str(user.id)
+      - id: remote_resources
+        if: user is not None
+        execute: |
+          user_preferences = user.extra_preferences
+          pulsar_tag = None
+          for data_item in user_preferences:
+            if "distributed_compute|remote_resources" in data_item:
+                if user_preferences[data_item] != "None":
+                    pulsar_tag = user_preferences[data_item]
+          new_tag = [{ 'require': pulsar_tag }] if pulsar_tag != None else []
+          entity.tpv_tags = entity.tpv_tags.combine(self.tpv_tags, TagSetManager.from_dict(new_tag or {}))
```

TagSetManager excepts `list[dict]` as input.

```suggestion
          entity.tpv_tags = entity.tpv_tags.combine(self.tpv_tags, TagSetManager.from_dict(new_tag))
```

---

Is this a bug then? https://github.com/galaxyproject/total-perspective-vortex/blob/6e912329c02c483685a37f6aa4485eeab027e095/tpv/core/entities.py#LL196C15-L196C64

---

Sorry the "bug" is actually [the type hint on entities.py](https://github.com/galaxyproject/total-perspective-vortex/blob/6e912329c02c483685a37f6aa4485eeab027e095/tpv/core/entities.py#L161), which is wrong.

It makes no sense to define the signature as `def from_dict(tags: list[dict]) -> TagSetManager:` but then do stuff like `tags.get('require')`. I trusted TPV's type hints too much.

I'll make a new changes request review with suggestions.